File: java/src/main/java/io/valkey/javabenchmark/config/DriverConfig.java
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.valkey.javabenchmark.config;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Map;

/**
 * Configuration model for driver settings.
 * Maps to the driver JSON configuration files.
 *
 * @author Ilia Kolominsky
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class DriverConfig {

    @JsonProperty("schema_version")
    private String schemaVersion;

    @JsonProperty("description")
    private String description;

    @JsonProperty("driver_id")
    private String driverId;

    @JsonProperty("mode")
    private String mode;

    @JsonProperty("auth")
    private AuthConfig auth;

    @JsonProperty("tls")
    private TlsConfig tls;

    @JsonProperty("specific_driver_config")
    private Map<String, Object> specificDriverConfig;

    // Getters and Setters

    public String getSchemaVersion() {
        return schemaVersion;
    }

    public void setSchemaVersion(String schemaVersion) {
        this.schemaVersion = schemaVersion;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getDriverId() {
        return driverId;
    }

    public void setDriverId(String driverId) {
        this.driverId = driverId;
    }

    public String getMode() {
        return mode;
    }

    public void setMode(String mode) {
        this.mode = mode;
    }

    public AuthConfig getAuth() {
        return auth;
    }

    public void setAuth(AuthConfig auth) {
        this.auth = auth;
    }

    public TlsConfig getTls() {
        return tls;
    }

    public void setTls(TlsConfig tls) {
        this.tls = tls;
    }

    public Map<String, Object> getSpecificDriverConfig() {
        return specificDriverConfig;
    }

    public void setSpecificDriverConfig(Map<String, Object> specificDriverConfig) {
        this.specificDriverConfig = specificDriverConfig;
    }

    // Convenience methods

    /**
     * Check if this is a cluster mode configuration.
     * 
     * @return true if mode is "cluster"
     */
    public boolean isClusterMode() {
        return "cluster".equalsIgnoreCase(mode);
    }

    /**
     * Check if TLS is enabled.
     * 
     * @return true if tls config is present
     */
    public boolean isTlsEnabled() {
        return tls != null;
    }

    /**
     * Check if authentication is configured.
     * 
     * @return true if auth config is present with password
     */
    public boolean hasAuth() {
        return auth != null && auth.getPassword() != null && !auth.getPassword().isEmpty();
    }

    /**
     * Check if connection pooling is enabled in specific_driver_config.
     * Used by Spring Data drivers with Jedis to enable JedisPool (matching Spring Boot defaults).
     * 
     * @return true if use_pooling is set to true in specific_driver_config
     */
    public boolean isUsePooling() {
        if (specificDriverConfig != null) {
            Object value = specificDriverConfig.get("use_pooling");
            if (value instanceof Boolean) return (Boolean) value;
            if (value instanceof String) return Boolean.parseBoolean((String) value);
        }
        return false;
    }

    /**
     * Get the connection pool size from specific_driver_config.
     * Used by Spring Data drivers (Jedis pool maxTotal, ValkeyGlide adapter pool maxPoolSize).
     * Defaults to 8, matching Spring Boot / Spring Data defaults.
     * 
     * @return pool size (default 8)
     */
    public int getPoolSize() {
        if (specificDriverConfig != null) {
            Object value = specificDriverConfig.get("pool_size");
            if (value instanceof Number) return ((Number) value).intValue();
            if (value instanceof String) {
                try { return Integer.parseInt((String) value); } catch (NumberFormatException ignored) {}
            }
        }
        return 8; // Spring Boot default
    }

    /**
     * Get the secondary driver ID for spring-data-* drivers.
     * 
     * @return secondary driver ID or null
     */
    public String getSecondaryDriverId() {
        if (specificDriverConfig != null) {
            Object secondaryId = specificDriverConfig.get("secondary_driver_id");
            // Also check for the typo version in the config
            if (secondaryId == null) {
                secondaryId = specificDriverConfig.get("scondary_driver_id");
            }
            return secondaryId != null ? secondaryId.toString() : null;
        }
        return null;
    }

    /**
     * Get the secondary driver configuration.
     * 
     * @return secondary driver config map or null
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> getSecondaryDriverConfig() {
        if (specificDriverConfig != null) {
            Object config = specificDriverConfig.get("secondary_driver_config");
            if (config instanceof Map) {
                return (Map<String, Object>) config;
            }
        }
        return null;
    }

    @Override
    public String toString() {
        return "DriverConfig{" +
                "driverId='" + driverId + '\'' +
                ", mode='" + mode + '\'' +
                ", tlsEnabled=" + isTlsEnabled() +
                ", hasAuth=" + hasAuth() +
                '}';
    }

    /**
     * Authentication configuration.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class AuthConfig {

        @JsonProperty("username")
        private String username;

        @JsonProperty("password")
        private String password;

        public String getUsername() {
            return username;
        }

        public void setUsername(String username) {
            this.username = username;
        }

        public String getPassword() {
            return password;
        }

        public void setPassword(String password) {
            this.password = password;
        }
    }

    /**
     * TLS configuration.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class TlsConfig {

        @JsonProperty("cert_path")
        private String certPath;

        @JsonProperty("key_path")
        private String keyPath;

        @JsonProperty("ca_cert_path")
        private String caCertPath;

        @JsonProperty("verify_peer")
        private Boolean verifyPeer;

        public String getCertPath() {
            return certPath;
        }

        public void setCertPath(String certPath) {
            this.certPath = certPath;
        }

        public String getKeyPath() {
            return keyPath;
        }

        public void setKeyPath(String keyPath) {
            this.keyPath = keyPath;
        }

        public String getCaCertPath() {
            return caCertPath;
        }

        public void setCaCertPath(String caCertPath) {
            this.caCertPath = caCertPath;
        }

        public Boolean getVerifyPeer() {
            return verifyPeer;
        }

        public void setVerifyPeer(Boolean verifyPeer) {
            this.verifyPeer = verifyPeer;
        }
    }
}

File: csharp/src/RespBench/Config/DriverConfig.cs
/*
 * Copyright 2025 the original author or authors.
 */
using System.Text.Json.Serialization;

namespace RespBench.Config;

/// <summary>
/// Configuration model for driver settings.
/// Maps to the driver JSON configuration files.
/// </summary>
public class DriverConfig
{
    [JsonPropertyName("schema_version")]
    public string? SchemaVersion { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("driver_id")]
    public string DriverId { get; set; } = "";

    [JsonPropertyName("mode")]
    public string Mode { get; set; } = "";

    [JsonPropertyName("auth")]
    public AuthConfig? Auth { get; set; }

    [JsonPropertyName("tls")]
    public TlsConfig? Tls { get; set; }

    [JsonPropertyName("command_timeout_ms")]
    public int? CommandTimeoutMs { get; set; }

    [JsonPropertyName("specific_driver_config")]
    public Dictionary<string, object>? SpecificDriverConfig { get; set; }

    // Convenience methods

    /// <summary>Check if this is a cluster mode configuration.</summary>
    public bool IsClusterMode => "cluster".Equals(Mode, StringComparison.OrdinalIgnoreCase);

    /// <summary>Check if TLS is enabled.</summary>
    public bool IsTlsEnabled => Tls != null;

    /// <summary>Check if authentication is configured.</summary>
    public bool HasAuth => Auth?.Password != null && !string.IsNullOrEmpty(Auth.Password);

    /// <summary>Check if connection pooling is enabled in specific_driver_config.</summary>
    public bool IsUsePooling => GetSpecificBool("use_pooling", false);

    /// <summary>Get connection pool size from specific_driver_config (default 8).</summary>
    public int PoolSize => GetSpecificInt("pool_size", 8);

    /// <summary>Get the secondary driver ID for composite drivers.</summary>
    public string? SecondaryDriverId
    {
        get
        {
            if (SpecificDriverConfig == null) return null;
            if (SpecificDriverConfig.TryGetValue("secondary_driver_id", out var value))
                return value?.ToString();
            if (SpecificDriverConfig.TryGetValue("scondary_driver_id", out value))
                return value?.ToString();
            return null;
        }
    }

    /// <summary>Get secondary driver configuration map.</summary>
    public Dictionary<string, object>? SecondaryDriverConfig
    {
        get
        {
            if (SpecificDriverConfig?.TryGetValue("secondary_driver_config", out var value) == true
                && value is Dictionary<string, object> dict)
                return dict;
            return null;
        }
    }

    private bool GetSpecificBool(string key, bool defaultValue)
    {
        if (SpecificDriverConfig?.TryGetValue(key, out var value) != true) return defaultValue;
        if (value is bool b) return b;
        if (value is System.Text.Json.JsonElement je && je.ValueKind == System.Text.Json.JsonValueKind.True) return true;
        if (value is System.Text.Json.JsonElement je2 && je2.ValueKind == System.Text.Json.JsonValueKind.False) return false;
        if (value is string s) return bool.TryParse(s, out var parsed) && parsed;
        return defaultValue;
    }

    private int GetSpecificInt(string key, int defaultValue)
    {
        if (SpecificDriverConfig?.TryGetValue(key, out var value) != true) return defaultValue;
        if (value is int i) return i;
        if (value is long l) return (int)l;
        if (value is double d) return (int)d;
        if (value is System.Text.Json.JsonElement je && je.TryGetInt32(out var jInt)) return jInt;
        if (value is string s && int.TryParse(s, out var parsed)) return parsed;
        return defaultValue;
    }

    internal double GetSpecificDouble(string key, double defaultValue)
    {
        if (SpecificDriverConfig?.TryGetValue(key, out var value) != true) return defaultValue;
        if (value is double d) return d;
        if (value is float f) return f;
        if (value is int i) return i;
        if (value is long l) return l;
        if (value is System.Text.Json.JsonElement je && je.TryGetDouble(out var jd)) return jd;
        if (value is string s && double.TryParse(s, out var parsed)) return parsed;
        return defaultValue;
    }

    internal long GetSpecificLong(string key, long defaultValue)
    {
        if (SpecificDriverConfig?.TryGetValue(key, out var value) != true) return defaultValue;
        if (value is long l) return l;
        if (value is int i) return i;
        if (value is double d) return (long)d;
        if (value is System.Text.Json.JsonElement je && je.TryGetInt64(out var jl)) return jl;
        if (value is string s && long.TryParse(s, out var parsed)) return parsed;
        return defaultValue;
    }

    internal string? GetSpecificString(string key)
    {
        if (SpecificDriverConfig?.TryGetValue(key, out var value) != true) return null;
        if (value is System.Text.Json.JsonElement je && je.ValueKind == System.Text.Json.JsonValueKind.String)
            return je.GetString();
        return value?.ToString();
    }

    public override string ToString() =>
        $"DriverConfig{{driverId='{DriverId}', mode='{Mode}', tlsEnabled={IsTlsEnabled}, hasAuth={HasAuth}}}";
}

/// <summary>Authentication configuration.</summary>
public class AuthConfig
{
    [JsonPropertyName("username")]
    public string? Username { get; set; }

    [JsonPropertyName("password")]
    public string? Password { get; set; }
}

/// <summary>TLS configuration.</summary>
public class TlsConfig
{
    [JsonPropertyName("cert_path")]
    public string? CertPath { get; set; }

    [JsonPropertyName("key_path")]
    public string? KeyPath { get; set; }

    [JsonPropertyName("ca_cert_path")]
    public string? CaCertPath { get; set; }

    [JsonPropertyName("verify_peer")]
    public bool? VerifyPeer { get; set; }
}
